ast/edit/testcase.go

```diff
@@ -1,60 +1,241 @@
 package edit
 
 import (
+	"context"
 	"fmt"
+	"path/filepath"
+	"strings"
 
 	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/dependencies/filesystem"
+	"github.com/influxdata/flux/parser"
 )
 
-/* TestCaseTransform is a macro that transforms ast that uses
-   the testcase statement into multiple independent packages with
-   a single flux script that is the test.
-
-   For instance, the following flux:
-
-import "testing"
-myVar = 4
-testcase addition { test.assertEqual(want: 2 + 2, got: myVar }
-
-	...would be transformed into the equivalent script...
-
-import "testing"
-myVar = 4
-test.assertEqual(want: 2 + 2, got: myVar
-*/
-func TestcaseTransform(pkg *ast.Package) ([]*ast.Package, error) {
+// TestCaseTransform will transform an *ast.Package into a set of *ast.Package values
+// that represent each testcase defined within the original package.
+//
+// A testcase is defined with the testcase statement such as below.
+//
+//     import "testing/assert"
+//     myVar = 4
+//     testcase addition {
+//         assert.equal(want: 2 + 2, got: myVar)
+//     }
+//
+// This gets transformed into a package that looks like this:
+//
+//     import "testing/assert"
+//     myVar = 4
+//     assert.equal(want: 2 + 2, got: myVar)
+//
+// It is allowed to include options within the testcase block as they will be extracted
+// to the top level.
+//
+// In addition to this syntax, testcase blocks may also include imports to other test files.
+// These imports will transform the imported testcase in a slightly different way.
+// If a testcase is included by import, it will be translated such as:
+//
+//     import "testing/assert"
+//     math_test = () => {
+//         myVar = 4
+//         test_addition = () => {
+//             assert.equal(want: 2 + 2, got: myVar)
+//             return {}
+//         }
+//         return {myVar, test_addition}
+//     }()
+//
+// Each import will be prepended to the list of files in the package as its own file.
+//
+// If a testcase includes an import statement, it will be replaced with the given body.
```

Comment:
Is this only for recursive imports? Or all imports?

Reply:
This is only if you try and import a testcase that itself has a testcase. We discussed this, and while it opens up a can of worms to support it recursively, allowing deeply nested test cases is something we'd rather actively discourage as bad practice.

Reply:
It's all imports at the moment. In the normal transformation of a testcase, the import gets extracted to the top level because that's just what testcase transforms do. This puts the syntax in a context where it can import or specify options. My original thought was to translate every testcase as:

```
testcase foo {
    bar()
}

foo = () => {
    bar()
}
foo()
```

This would make the two transformations identical. But, it means that the testcase body can't support options since options cannot be specified outside of the top level. Since a primary reason for this feature is to set planner rules through options, that was a pretty hard requirement.

But, since imports have to be translated this way, we cannot fully support options or imports without making some guesses based on syntax that the AST level really doesn't know. The AST may be able to see `option now = <expr>`, but it cannot make any assumptions about what `now` is because only the interpreter knows what that value really is. The reason we need to translate these into functions is because we don't know which one will be invoked by the testcase when it is imported because that would require the support of the interpreter which we don't want to do at this moment.

My thoughts are that we will need to refactor testcases into a formal part of the interpreter or execution engine when we want to release this feature for external users. In addition to that, we will need to fully define how we intend the package system to be used and formally support user-defined imports.

I think we would be able to make some different choices in the design with those two features implemented.

For one, we wouldn't have to perform any translation of an import into a function and would not have to worry about potentially conflicting identifiers. The package scoping system would prevent that and options can be specified multiple times across different packages. But, options cannot be specified multiple times in the same package which limits us in this circumstance.

Second, if the execution engine/interpreter were executing the test cases, we would not have to perform AST translations which means that we would have more rich knowledge of what an identifier references rather than guessing what an identifier might be referencing. This point is more minor compared to the previous point about user-defined imports.